addons/mail/models/mail_thread.py

```diff
@@ -2008,112 +2022,226 @@ def message_post(self, *,
-            all_attachments = self.env['ir.attachment'].browse([attachment_tuple[1] for attachment_tuple in attachment_ids])
-            prioritary_attachments = all_attachments.filtered(lambda x: x.mimetype.endswith('pdf')) \
-                                     or all_attachments.filtered(lambda x: x.mimetype.startswith('image')) \
-                                     or all_attachments
-            self.sudo().with_context(tracking_disable=True).write({'message_main_attachment_id': prioritary_attachments[0].id})
-
     def _message_post_after_hook(self, message, msg_vals):
         """ Hook to add custom behavior after having posted the message. Both
         message and computed value are given, to try to lessen query count by
         using already-computed values instead of having to rebrowse things. """
+        # Set main attachment field if necessary
+        self._message_set_main_attachment_id([att[1] for att in msg_vals['attachment_ids'] or []])
+
+    def _message_mail_after_hook(self, mails):
+        """ Hook to add custom behavior after having sent an mass mailing.
+
+        :param mail.mail mails: mail.mail records about to be sent"""
+        return
+
+    def _message_set_main_attachment_id(self, attachment_ids):
+        if self._name == 'mail.thread':
+            return
+        if attachment_ids and not self.message_main_attachment_id:
+            all_attachments = self.env['ir.attachment'].browse(attachment_ids)
+            prioritary_attachments = all_attachments.filtered(lambda x: x.mimetype.endswith('pdf')) \
+                                     or all_attachments.filtered(lambda x: x.mimetype.startswith('image')) \
+                                     or all_attachments
+            self.sudo().with_context(tracking_disable=True).message_main_attachment_id = prioritary_attachments[0].id
 
     # ------------------------------------------------------------
     # MESSAGE POST API / WRAPPERS
     # ------------------------------------------------------------
 
-    def _message_compose_with_view(self, views_or_xmlid, message_log=False, **kwargs):
-        """ Helper method to send a mail / post a message / log a note using
-        a view_id to render using the ir.qweb engine. This method is stand
-        alone, because there is nothing in template and composer that allows
-        to handle views in batch. This method should probably disappear when
-        templates handle ir ui views. """
-        values = kwargs.pop('values', None) or dict()
-        try:
-            from odoo.addons.http_routing.models.ir_http import slug
-            values['slug'] = slug
-        except ImportError:
-            values['slug'] = lambda self: self.id
-        view_ref = views_or_xmlid.id if isinstance(views_or_xmlid, models.BaseModel) else views_or_xmlid
+    def message_mail_with_source(self, template_id=False, view_ref=False, render_values=None,
```

Comment:
Because now we have either a template or a view, we can't use `message_mail_with_source` with positionnal argument
```
message_mail_with_source(x, y, z)
vs
message_mail_with_source(template_id=x, render_values=...)
```

Can't we have only 1 argument for both `source`, and we check if it's an integer, then the `._name` to know if it's a template or a view ?

Also it will make view vs template consistent (currently you can give a recordset or integer for the view, not for the template)

What do you think ?

Reply:
I was wondering the same thing then went for template_id or source_ref. But indeed I prefer to have a positionnal argument. So back to the unique source with some checks: should be a record or an xml id (obviously an ID we cannot do anything with it).